Refactor `backend.js` #5 - Simplify the main logic

nemanjaglumac opened the pull request:
### Status
PENDING CI && PENDING REVIEW

### What does this PR accomplish?
- This is the fifth step/task in https://github.com/metabase/metabase/issues/24023
- Greatly simplifies the previous logic for `BackendResource` whose only job was to spin up (temporary) backend for Cypress

> **Note**
> I tried to keep changes logically contained to a single commit. That's probably the easiest way to review this PR, rather than trying to look at the final diff.


As before, everything should still work the same as before when you spin Cypress up (both locally and in the CI).

### Questions:
What do you think about simplifying this even further and dealing with three separate functions that we can export from `cypress-runner-backend.js`?

Something like:
```js
// cypress-runner-backend.js

module.exports = { createServer, startBackend, stopBackend }
```

and then using them in `run_cypress_tests.js` like so:
```js
const server = createServer(/*we can pass options for the server directly here*/)

startBakend(server, /* also possible to pass even more additional options, or flags for the backend if needed */)

// and in the end
stopBackend(server)
```

### Additional info:
- If we decide to go down this route, then this PR would replace https://github.com/metabase/metabase/pull/24077 created by @ariya as an alternative approach
- I didn't compare the times explicitly, but it seems like backend is now spawning MUCH faster

nemanjaglumac committed:
Simplify temporary DB path

nemanjaglumac committed:
Set the port to 4000

nemanjaglumac committed:
Move logic to create server config inside the base function

nemanjaglumac committed:
Rename `kill` function to distinguish between it and Node process `kill`

nemanjaglumac committed:
`createServer` as a method

nemanjaglumac committed:
Remove reliance on Maps

nemanjaglumac committed:
Stop server logic implemented directly in a method

nemanjaglumac committed:
Remove unneeded parameter

nemanjaglumac committed:
Start server logic implemented directly in a method

nemanjaglumac committed:
Remove `isReady` from global scope

nemanjaglumac committed:
Remove `delay` from the global scope

nemanjaglumac committed:
Rename `BackendResource` to `CypressBackend`

nemanjaglumac committed:
Rename file

gusaiani commented:
@nemanjaglumac 

> module.exports = { createServer, startBackend, stopBackend }

Yes yes yes 👍

gusaiani reviewed (approved):
Looking better and better.

npfitz commented on frontend/test/__runner__/cypress-runner-backend.js:
```diff
@@ -0,0 +1,151 @@
+
+const CypressBackend = e2eBackend();
+
+function e2eBackend() {
```
I think we can remove a layer here and just have `CypressBackend` set to what `e2eBackend()`, without actually calling `e2eBackend`
```suggestion
const CypressBackend = {
 createServer() {...},
 start(server) {...},
 ...
}
```

npfitz replied:
Haha, @gusaiani beat me to it

nemanjaglumac replied:
@npfitz done! Thank you for the suggestion.

nemanjaglumac committed:
Simplify `CypressBackend`

No need to invoke the function when we can store everything
as method definitions under `CypressBackend` object.

nemanjaglumac committed:
Pass `port` as an argument with default value

nemanjaglumac committed:
Use template literals for logging

nemanjaglumac committed:
Destructure `response`

nemanjaglumac committed:
Fix small typo in the comment

nemanjaglumac committed:
Tidy up `dbFile` in the `server`
